File: 2023/cpp/day1/main.cc

#include <cctype>
#include <cstdio>
#include <fstream>
#include <iostream>
#include <string>
#include <string_view>
#include <utility>
const std::pair<std::string, int> numbers[9] = {
    {"one", 1}, {"two", 2},   {"three", 3}, {"four", 4}, {"five", 5},
    {"six", 6}, {"seven", 7}, {"eight", 8}, {"nine", 9}};

int match_number(std::string_view view) {
  for (const auto &[word, number] : numbers) {
    if (view.rfind(word, 0) == 0) {
      return number;
    }
  }
  return -1;
}

int main() {
  const auto file_name = "phase1.txt";
  std::ifstream file(file_name);
  std::string line;

  int sum = 0;

  if (file.is_open()) {
    while (std::getline(file, line)) {
      int first_num = -1;
      int last_num = 0;

      for (int i = 0; i < line.size(); ++i) {
        char c = line[i];
        if (std::isdigit(c)) {
          if (first_num == -1) {
            first_num = c - '0';
          }
          last_num = c - '0';
        } else {
          int number = match_number(line.substr(i, line.size() - 1));
          if (number != -1) {
            if (first_num == -1) {
              first_num = number;
            }
            last_num = number;
          }
        }
      }

      sum += first_num * 10 + last_num;
    }
  }

  std::cout << sum << std::endl;

  return 0;
}

File: 2023/day1/src/main.rs
use std::{io::{self, BufRead, BufReader, Lines}, fs::File, path::Path};


fn read_lines<P> (path: P) -> io::Result<Lines<BufReader<File>>>
where P: AsRef<Path>, {
    let file = File::open(path)?;
    Ok(BufReader::new(file).lines())
}

fn match_number(input: &str) -> Option<u8> {
    let number_words = [
        ("one", 1),
        ("two", 2),
        ("three", 3),
        ("four", 4),
        ("five", 5),
        ("six", 6),
        ("seven", 7),
        ("eight", 8),
        ("nine", 9)
    ];

    for (word, number) in number_words.iter() {
        if input.starts_with(word) {
            return Some(*number);
        }
    }

    None
}

fn main() {
    let lines = match read_lines("phase1.txt") {
        Ok(lines) => lines,
        Err(e) => panic!("Error: {}", e),
    };

    let mut sum = 0;

    for line in lines {
        if let Ok(s) = line {
            let mut first_num: u32 = 10;
            let mut last_num: u32 = 0;
            for i in 0..s.len() {
                let c = s.chars().nth(i).unwrap();

                if c.is_numeric() {
                    if first_num == 10 {
                        first_num = c.to_digit(10).unwrap();
                    }
                    last_num = c.to_digit(10).unwrap();
                } else {
                    let parsed_num = match_number(&s[i..]);
                    if let Some(num) = parsed_num {
                        if first_num == 10 {
                            first_num = num as u32;
                        }
                        last_num = num as u32;
                    }
                }
            }

            println!("{} {} {}", s, first_num, last_num);

            sum += (first_num * 10) + last_num;
        }
    }

    println!("Sum: {}", sum);
}
